internal/caveats/diff.go

```diff
@@ -0,0 +1,138 @@
+package caveats
+
+import (
+	"golang.org/x/exp/maps"
+
+	"github.com/authzed/spicedb/internal/util"
+	"github.com/authzed/spicedb/pkg/caveats/types"
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+)
+
+// DeltaType defines the type of caveat deltas.
+type DeltaType string
+
+const (
+	// CaveatAdded indicates that the caveat was newly added/created.
+	CaveatAdded DeltaType = "caveat-added"
+
+	// CaveatRemoved indicates that the caveat was removed.
+	CaveatRemoved DeltaType = "caveat-removed"
+
+	// AddedParameter indicates that the parameter was added to the caveat.
+	AddedParameter DeltaType = "added-parameter"
+
+	// RemovedParameter indicates that the parameter was removed from the caveat.
+	RemovedParameter DeltaType = "removed-parameter"
+
+	// ParameterTypeChanged indicates that the type of the parameter was changed.
+	ParameterTypeChanged DeltaType = "parameter-type-changed"
+)
+
+// Diff holds the diff between two caveats.
+type Diff struct {
+	existing *core.CaveatDefinition
+	updated  *core.CaveatDefinition
+	deltas   []Delta
+}
+
+// Deltas returns the deltas between the two caveats.
+func (cd Diff) Deltas() []Delta {
+	return cd.deltas
+}
+
+// Delta holds a single change of a caveat.
+type Delta struct {
```

Comment:
This exactly follows the format in the Namespace Diff library. It was done this way because it makes it *very* easy to switch over the type of diff in the schema comparison code. Otherwise, you need a DeltaType of "CaveatUnchanged" and then have to check the parameters' diff too, which is not as nice IMO